components/cosmetic_filters/renderer/cosmetic_filters_js_handler.cc

```diff
@@ -0,0 +1,374 @@
+    const std::string& input) {
+  if (!EnsureConnected())
+    return;
+
+  cosmetic_filters_resources_->HiddenClassIdSelectors(
+      input, exceptions_,
+      base::BindOnce(&CosmeticFiltersJSHandler::OnHiddenClassIdSelectors,
+                     base::Unretained(this)));
+}
+
+void CosmeticFiltersJSHandler::AddJavaScriptObjectToFrame(
+    v8::Local<v8::Context> context) {
+  v8::Isolate* isolate = blink::MainThreadIsolate();
+  v8::HandleScope handle_scope(isolate);
+  if (context.IsEmpty())
+    return;
+
+  v8::Context::Scope context_scope(context);
+
+  CreateWorkerObject(isolate, context);
+}
+
+void CosmeticFiltersJSHandler::CreateWorkerObject(
+    v8::Isolate* isolate,
+    v8::Local<v8::Context> context) {
+  v8::Local<v8::Object> global = context->Global();
+  v8::Local<v8::Object> cosmetic_filters_obj;
+  v8::Local<v8::Value> cosmetic_filters_value;
+  if (!global->Get(context, gin::StringToV8(isolate, "cf_worker"))
+           .ToLocal(&cosmetic_filters_value) ||
+      !cosmetic_filters_value->IsObject()) {
+    cosmetic_filters_obj = v8::Object::New(isolate);
+    global
+        ->Set(context, gin::StringToSymbol(isolate, "cf_worker"),
+              cosmetic_filters_obj)
+        .Check();
+    BindFunctionsToObject(isolate, cosmetic_filters_obj);
+  }
+}
+
+void CosmeticFiltersJSHandler::BindFunctionsToObject(
+    v8::Isolate* isolate,
+    v8::Local<v8::Object> javascript_object) {
+  BindFunctionToObject(
+      isolate, javascript_object, "hiddenClassIdSelectors",
+      base::BindRepeating(&CosmeticFiltersJSHandler::HiddenClassIdSelectors,
+                          base::Unretained(this)));
+}
+
+template <typename Sig>
+void CosmeticFiltersJSHandler::BindFunctionToObject(
+    v8::Isolate* isolate,
+    v8::Local<v8::Object> javascript_object,
+    const std::string& name,
+    const base::RepeatingCallback<Sig>& callback) {
+  v8::Local<v8::Context> context = isolate->GetCurrentContext();
+  // Get the isolate associated with this object.
+  javascript_object
+      ->Set(context, gin::StringToSymbol(isolate, name),
+            gin::CreateFunctionTemplate(isolate, callback)
+                ->GetFunction(context)
+                .ToLocalChecked())
+      .Check();
+}
+
+bool CosmeticFiltersJSHandler::EnsureConnected() {
+  if (!cosmetic_filters_resources_.is_bound()) {
+    render_frame_->GetBrowserInterfaceBroker()->GetInterface(
+        cosmetic_filters_resources_.BindNewPipeAndPassReceiver());
+  }
+
+  if (cosmetic_filters_resources_.is_bound())
+    return true;
+  else
+    return false;
+}
+
+void CosmeticFiltersJSHandler::ProcessURL(const GURL& url) {
+  if (!EnsureConnected())
+    return;
+
+  url_ = url;
+  cosmetic_filters_resources_->ShouldDoCosmeticFiltering(
+      url_.spec(),
+      base::BindOnce(&CosmeticFiltersJSHandler::OnShouldDoCosmeticFiltering,
+                     base::Unretained(this)));
+}
+
+void CosmeticFiltersJSHandler::OnShouldDoCosmeticFiltering(
```

Comment:
do we need this hop (aka `OnShouldDoCosmeticFiltering`)? I. e. can we just check `!enabled`  on the browser side and immediately do `UrlCosmeticResources`, then callback to `OnUrlCosmeticResources`

Reply:
we've discussed that with @bridiver and came to the conclusion to use it like that. The problem with the solution you propose is that we need one more boolean here `enabled_1st_party_cf_filtering_` and we need to store it to use in `CosmeticFiltersJSHandler` calls. We could potentially return it with `OnUrlCosmeticResources` as well, but we've decided to use it like it is right now.

Reply:
okay, nothing critical, but I think it worth doing at some point